File: Top_100_liked_questions/Q1_E/Solution.cpp
#include <string>;
#include <stack>;
using namespace std;
class Solution {

public:bool isValid(string s) {
    // O(n) space time solution 
        if(s.size()%2 != 0 ) return false;
        stack<char> stk;
        for(char ch: s)
        {
            switch(ch)
            {
                case '(':stk.push(')');break;
                case '{':stk.push('}');break;
                case '[': stk.push(']');break;
                default: if(stk.empty() || stk.top() != ch) return false;
                    else stk.pop();
            }
        }
        
        return stk.empty();
    }
};

File: Top_100_liked_questions/Q1_E/Solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        dict = {'(':')','{':'}','[':']',};
        stack :list= []
        for ch in s:
            if(dict.get(ch,None) != None):
                stack.append()(dict[ch])
            elif (len(stack) == 0 or ch != stack.pop() ):
                return False;
        else:
            return  len(stack) ==0;

    def isValid_Optimal(self,s:str) -> bool :
        dict = {'(':')','{':'}','[':']',};
        stk = [];
        for ch in s:
            if ch in dict:
                stk.append(dict[ch]);
            elif not stk or ch is not stk.pop():
                return False;

        return len(stk) == 0